edge-modules/TwinTester/TwinOperator.cs

```diff
@@ -0,0 +1,354 @@
+            }
+        }
+
+        private async Task PerformDesiredPropertyUpdate()
+        {
+            try
+            {
+                string desiredPropertyUpdate = new string('1', Settings.Current.TwinUpdateCharCount); // dummy twin update needs to be any number
+                string patch = string.Format("{{ properties: {{ desired: {{ {0}: {1}}} }} }}", this.twinState.DesiredPropertyUpdateCounter, desiredPropertyUpdate);
+                Twin newTwin = await this.registryManager.UpdateTwinAsync(Settings.Current.DeviceId, Settings.Current.ModuleId, patch, this.twinState.TwinETag);
+                this.twinState.TwinETag = newTwin.ETag;
+                Logger.LogInformation($"Made desired property update {this.twinState.DesiredPropertyUpdateCounter}");
+            }
+            catch (Exception e)
+            {
+                Logger.LogInformation($"Failed call to desired property update: {e}");
+                return;
+            }
+
+            try
+            {
+                await this.storage.AddDesiredPropertyUpdate(this.twinState.DesiredPropertyUpdateCounter.ToString());
+                this.twinState.DesiredPropertyUpdateCounter += 1;
+            }
+            catch (Exception e)
+            {
+                Logger.LogError($"Failed adding desired property update to storage: {e}");
+            }
+        }
+
+        private async Task PerformReportedPropertyUpdate()
+        {
+            string reportedPropertyUpdate = new string('1', Settings.Current.TwinUpdateCharCount); // dummy twin update needs to be any number
+            var twin = new TwinCollection();
+            twin[this.twinState.ReportedPropertyUpdateCounter.ToString()] = reportedPropertyUpdate;
+            try
+            {
+                await this.moduleClient.UpdateReportedPropertiesAsync(twin);
+                Logger.LogInformation($"Made reported property update {this.twinState.ReportedPropertyUpdateCounter}");
+            }
+            catch (Exception e)
+            {
+                string failureStatus = $"{(int)StatusCode.ReportedPropertyUpdateCallFailure}: Failed call to update reported properties";
+                Logger.LogError(failureStatus + $": {e}");
+                await this.CallAnalyzerToReportStatus(Settings.Current.ModuleId, failureStatus, string.Empty);
+                return;
+            }
+
+            try
+            {
+                await this.storage.AddReportedPropertyUpdate(this.twinState.ReportedPropertyUpdateCounter.ToString());
+                this.twinState.ReportedPropertyUpdateCounter += 1;
+            }
+            catch (Exception e)
+            {
+                Logger.LogError($"Failed adding reported property update to storage: {e}");
+                return;
+            }
+        }
+
+        public async Task PerformUpdates()
+        {
+            lock (this.operationLock)
+            {
+                this.PerformDesiredPropertyUpdate().Wait();
```

Comment:
Don't use Wait() or Result anywhere.. use async/await instead

Reply:
I tried to use await here, but could not use inside the lock. I switched to using semaphore slim instead of lock.